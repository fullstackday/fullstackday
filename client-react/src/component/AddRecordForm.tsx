import React from "react";
import { Button, Container } from "@mui/material";
import { useForm } from "react-hook-form";
import { TextInput } from "./Input/TextInput";
import { DateTimeInput } from "./Input/DateTimeInput";
import { TimeRecord } from "../domain/TimeRecord";

interface Props {
  onSubmit: (values: Omit<TimeRecord, "_id">) => void;
  isLoading: boolean;
  record?: TimeRecord;
}

interface FormValues {
  date: string;
  start: string;
  end: string;
  project: string;
  comment?: string;
}

const recordToFormValues = (record: TimeRecord): FormValues => ({
  ...record,
  date: ymdString(record.start),
  start: record.start.toLocaleTimeString(),
  end: record.end.toLocaleTimeString(),
});

const formValuesToRecord = (values: FormValues): Omit<TimeRecord, "_id"> => ({
  ...values,
  start: new Date(values.date + "T" + values.start),
  end: new Date(values.date + "T" + values.end),
});

const ymdString = (date: Date): string =>
  date.getFullYear() +
  "-" +
  ("0" + (date.getMonth() + 1)).slice(-2) +
  "-" +
  ("0" + date.getDate()).slice(-2);

export const AddRecordForm: React.FunctionComponent<Props> = ({
  onSubmit,
  isLoading,
  record,
}) => {
  const {
    handleSubmit,
    control,
    formState: { errors, isValid },
  } = useForm({
    mode: "onChange",
    defaultValues: record
      ? recordToFormValues(record)
      : {
          project: "",
          comment: "",
          date: ymdString(new Date()),
          start: new Date().toLocaleTimeString(),
          end: "",
        },
  });

  return (
    <Container
      component="form"
      onSubmit={handleSubmit((values) => {
        onSubmit(formValuesToRecord(values));
      })}
      sx={{
        marginTop: 8,
        display: "flex",
        flexDirection: "column",
        alignItems: "center",
        "& .MuiTextField-root": {
          margin: "8px 0",
        },
      }}
    >
      <TextInput
        control={control}
        name="project"
        label="Project"
        rules={{ required: true }}
      />

      <TextInput
        control={control}
        name="date"
        type="date"
        label="Date"
        rules={{ required: true }}
      />
      <TextInput
        control={control}
        name="start"
        type="time"
        label="Start"
        rules={{ required: true }}
      />
      <TextInput
        control={control}
        name="end"
        type="time"
        label="End"
        rules={{ required: true }}
      />

      <TextInput
        control={control}
        name="comment"
        label="Comment"
        rules={{ required: false }}
      />

      <Button type="submit" disabled={!isValid || isLoading}>
        Create
      </Button>
    </Container>
  );
};
