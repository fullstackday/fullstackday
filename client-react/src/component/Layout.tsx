import React from "react";
import AppBar from "@mui/material/AppBar";
import Box from "@mui/material/Box";
import Toolbar from "@mui/material/Toolbar";
import Typography from "@mui/material/Typography";
import AdapterDateFns from "@mui/lab/AdapterDateFns";
import { LocalizationProvider } from "@mui/lab";
import CssBaseline from "@mui/material/CssBaseline";

interface Props {}

export const Layout: React.FunctionComponent<Props> = ({ children }) => {
  return (
    <LocalizationProvider dateAdapter={AdapterDateFns}>
      <Box sx={{ flexGrow: 1 }}>
        <CssBaseline />
        <AppBar position="static">
          <Toolbar>
            <Typography variant="h6" component="div" sx={{ flexGrow: 1 }}>
              WorkSchedule
            </Typography>
          </Toolbar>
        </AppBar>
        {children}
      </Box>
    </LocalizationProvider>
  );
};
