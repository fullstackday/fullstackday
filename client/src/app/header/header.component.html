<mat-toolbar color="primary">
    <mat-toolbar-row>
        <button mat-button matTooltip="Work Schedule" routerLink="/">
            WorkSchedule
        </button>
    </mat-toolbar-row>
</mat-toolbar>
