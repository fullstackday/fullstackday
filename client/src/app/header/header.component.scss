@import '~@angular/material/theming';
@include mat-core();

  $primary: mat-palette($mat-blue,200);
  $accent: mat-palette($mat-orange,200);

  $theme: mat-light-theme($primary, $accent);

@include angular-material-theme($theme);