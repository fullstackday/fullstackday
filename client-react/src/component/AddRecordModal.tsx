import React from "react";
import Box from "@mui/material/Box";
import Button from "@mui/material/Button";
import Typography from "@mui/material/Typography";
import Modal from "@mui/material/Modal";
import { AddRecordForm } from "./AddRecordForm";
import { LinearProgress } from "@mui/material";
import { useAddRecord } from "../hook/useAddRecord";

const style = {
  position: "absolute" as "absolute",
  top: "50%",
  left: "50%",
  transform: "translate(-50%, -50%)",
  width: 400,
  bgcolor: "background.paper",
  boxShadow: 24,
};

interface Props {}

export const AddRecordModal: React.FunctionComponent<Props> = () => {
  const [open, setOpen] = React.useState(false);
  const handleOpen = () => setOpen(true);
  const handleClose = () => setOpen(false);

  const { isLoading, addRecord } = useAddRecord({
    onSuccess: () => {
      handleClose();
    },
  });

  return (
    <>
      <Button onClick={handleOpen}>Track Work</Button>
      <Modal
        open={open}
        onClose={handleClose}
        aria-labelledby="modal-modal-title"
        aria-describedby="modal-modal-description"
      >
        <Box>
          <Box sx={style}>
            <Box sx={{ p: 4 }}>
              <Typography id="modal-modal-title" variant="h6" component="h2">
                Create a new Record
              </Typography>
              <AddRecordForm onSubmit={addRecord} isLoading={isLoading} />
            </Box>
            {isLoading && <LinearProgress />}
          </Box>
        </Box>
      </Modal>
    </>
  );
};
