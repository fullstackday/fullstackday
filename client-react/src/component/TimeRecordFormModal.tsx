import React from "react";
import Modal from "@mui/material/Modal";
import Box from "@mui/material/Box";
import Typography from "@mui/material/Typography";
import { AddRecordForm, AddRecordFormValues } from "./AddRecordForm";
import { LinearProgress } from "@mui/material";
import { TimeRecord } from "../domain/TimeRecord";

interface Props {
  open: boolean;
  isLoading: boolean;
  onSubmit: (record: AddRecordFormValues) => void;
  onClose: () => void;
  record?: TimeRecord;
}

const style = {
  position: "absolute" as "absolute",
  top: "50%",
  left: "50%",
  transform: "translate(-50%, -50%)",
  width: 400,
  bgcolor: "background.paper",
  boxShadow: 24,
};

export const TimeRecordFormModal: React.FunctionComponent<Props> = ({
  onClose,
  open,
  isLoading,
  onSubmit,
  record,
}) => {
  return (
    <Modal open={open} onClose={onClose}>
      <Box>
        <Box sx={style}>
          <Box sx={{ p: 4 }}>
            <Typography id="modal-modal-title" variant="h6" component="h2">
              Create a new Record
            </Typography>
            <AddRecordForm
              onSubmit={onSubmit}
              isLoading={isLoading}
              record={record}
            />
          </Box>
          {isLoading && <LinearProgress />}
        </Box>
      </Box>
    </Modal>
  );
};
