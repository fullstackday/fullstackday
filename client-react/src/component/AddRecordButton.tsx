import React from "react";
import Box from "@mui/material/Box";
import Button from "@mui/material/Button";
import Typography from "@mui/material/Typography";
import Modal from "@mui/material/Modal";
import { AddRecordForm } from "./AddRecordForm";
import { LinearProgress } from "@mui/material";
import { useAddRecord } from "../hook/useAddRecord";
import { TimeRecordFormModal } from "./TimeRecordFormModal";

const style = {
  position: "absolute" as "absolute",
  top: "50%",
  left: "50%",
  transform: "translate(-50%, -50%)",
  width: 400,
  bgcolor: "background.paper",
  boxShadow: 24,
};

interface Props {}

export const AddRecordButton: React.FunctionComponent<Props> = () => {
  const [open, setOpen] = React.useState(false);
  const handleOpen = () => setOpen(true);
  const handleClose = () => setOpen(false);

  const { isLoading, addRecord } = useAddRecord({
    onSuccess: () => {
      handleClose();
    },
  });

  return (
    <>
      <Button onClick={handleOpen}>Track Work</Button>
      <TimeRecordFormModal
        open={open}
        isLoading={isLoading}
        onSubmit={addRecord}
        onClose={handleClose}
      />
    </>
  );
};
