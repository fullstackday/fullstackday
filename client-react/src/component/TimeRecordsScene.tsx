import React, { useState } from "react";
import { TimeRecordsTable } from "./TimeRecordsTable";
import { AddRecordButton } from "./AddRecordButton";
import { useGetAllRecords } from "../hook/useGetAllRecords";
import { useDeleteRecord } from "../hook/useDeleteRecord";
import { LinearProgress } from "@mui/material";
import { TimeRecordFormModal } from "./TimeRecordFormModal";
import { TimeRecord } from "../domain/TimeRecord";
import { useUpdateRecord } from "../hook/useUpdateRecord";

interface Props {}

export const TimeRecordsScene: React.FunctionComponent<Props> = () => {
  const [editRecord, setEditRecord] = useState<TimeRecord | undefined>();
  const { data, isLoading } = useGetAllRecords();
  const { deleteRecord, isLoading: isLoadingDelete } = useDeleteRecord();
  const { updateRecord, isLoading: isLoadingUpdate } = useUpdateRecord({
    onSuccess: () => setEditRecord(undefined),
  });

  if (!data) {
    return <div>Loadingâ€¦</div>;
  }

  return (
    <>
      {(isLoading || isLoadingDelete) && <LinearProgress />}
      <AddRecordButton />

      {editRecord && (
        <TimeRecordFormModal
          open={true}
          isLoading={isLoadingUpdate}
          onSubmit={(formValues) =>
            updateRecord({ record: editRecord, formValues })
          }
          onClose={() => setEditRecord(undefined)}
          record={editRecord}
        />
      )}
      <TimeRecordsTable
        timeRecords={data}
        onDeleteRecord={deleteRecord}
        onEditRecord={(record) => setEditRecord(record)}
      />
    </>
  );
};
