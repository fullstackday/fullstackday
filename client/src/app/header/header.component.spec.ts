import { ComponentFixture, TestBed } from '@angular/core/testing';

import { HeaderComponent } from './header.component';
import { MockModule } from 'ng-mocks';
import { MatToolbarModule } from '@angular/material/toolbar';

describe('HeaderComponent', () => {
    let component: HeaderComponent;
    let fixture: ComponentFixture<HeaderComponent>;

    beforeEach(async () => {
        await TestBed.configureTestingModule({
            declarations: [
                HeaderComponent
            ],
            imports: [
                MockModule(MatToolbarModule)
            ]
        })
            .compileComponents();
    });

    beforeEach(() => {
        fixture = TestBed.createComponent(HeaderComponent);
        component = fixture.componentInstance;
        fixture.detectChanges();
    });

    test('should match snapshot', () => {
        expect(fixture).toMatchSnapshot();
    });
});
