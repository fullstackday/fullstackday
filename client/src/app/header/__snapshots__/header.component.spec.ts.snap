// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HeaderComponent should match snapshot 1`] = `
<app-header
  __ngContext__={[Function LRootView]}
>
  <mat-toolbar
    color="primary"
  >
    <mat-toolbar-row>
      <button
        data-testid="home-button"
        mat-button=""
        mattooltip="Work Schedule"
        routerlink="/"
      >
         WorkSchedule 
      </button>
    </mat-toolbar-row>
  </mat-toolbar>
</app-header>
`;
